src/relay/dnsrelay/mod.rs

```diff
@@ -1,194 +1,90 @@
-    let mut name = qname.to_ascii();
-    name.pop();
-    let addr = Address::DomainNameAddress(name, 0);
-    let qname_in_proxy_list = context.check_qname_in_proxy_list(&addr);
-
-    let remote_response_fut = async {
-        match qname_in_proxy_list {
-            Some(false) => None,
-            _ => {
-                let timeout = Some(Duration::new(3, 0));
-                try_timeout(proxy_lookup(context.clone(), svr_cfg, remote, qname, qtype), timeout)
-                    .await
-                    .ok()
-            }
-        }
-    };
-
-    let local_response = match qname_in_proxy_list {
-        Some(true) => None,
-        _ => {
-            let timeout = Some(Duration::new(3, 0));
-            try_timeout(local_lookup(qname, qtype, local), timeout).await.ok()
-        }
-    }
-    .unwrap_or_else(Message::new);
+    let remote_response_fut = try_timeout(remote.lookup(query), Some(Duration::new(3, 0)));
+    let local_response_fut = try_timeout(local.lookup(query), Some(Duration::new(3, 0)));
 
-    match qname_in_proxy_list {
+    match context.check_query_in_proxy_list(query) {
         Some(true) => {
-            let remote_response = remote_response_fut.await.unwrap_or_else(Message::new);
+            let remote_response = remote_response_fut.await;
             debug!("pick remote response (qname): {:?}", remote_response);
-            return Ok((remote_response, true));
+            return (remote_response, true);
         }
         Some(false) => {
+            let local_response = local_response_fut.await;
             debug!("pick local response (qname): {:?}", local_response);
-            return Ok((local_response, false));
+            return (local_response, false);
         }
         None => (),
     }
 
-    if local_response.answer_count() == 0 {
-        let remote_response = remote_response_fut.await.unwrap_or_else(Message::new);
-        return Ok((remote_response, true));
-    }
-
-    for rec in local_response.answers() {
+    // FIXME: spawn(remote_response_fut)
+    let local_response = local_response_fut.await;
+    for rec in local_response.unwrap_or_else(|_| Message::new()).answers() {
+        if rec.record_type() != query.query_type() {
+            warn!("local DNS response has inconsistent answer type {} for query {}", rec.record_type(), query);
+            break
+        }
         let forward = match rec.rdata() {
             RData::A(ref ip) => context.check_ip_in_proxy_list(&IpAddr::from(*ip)),
             RData::AAAA(ref ip) => context.check_ip_in_proxy_list(&IpAddr::from(*ip)),
-            _ => true,
+            RData::PTR(_) => panic!("PTR records should not reach here"),
+            _ => context.is_default_in_proxy_list(),
         };
         if !forward {
-            debug!("pick local response (ip): {:?}", local_response);
-            return Ok((local_response, false));
+            debug!("pick local response (response): {:?}", local_response);
```

Comment:
Value borrowed here after move here:

https://github.com/shadowsocks/shadowsocks-rust/pull/242/files#diff-0fe08b50f6647bbea91fb44106ea4d16R68

Reply:
I didn't get a compile error though.

Reply:
Okay I see it now.